<img src='https://user-images.githubusercontent.com/6457691/89746244-a6e2a200-daf3-11ea-9314-7863d544d73c.png' width = '200' align = 'right'>

## *DATA SCIENCE / SECTION 1 / SPRINT 1 / NOTE 2*

# 📝 Assignment


## 1. Feature Engineering

- `KT&G` 기업 데이터를 계속 사용합니다.
  - 변수 이름은 `df`로 저장하세요.

- 각 분기별 영업이익률을 직접 계산해보세요. 
  - Feature 이름은 `영업이익률2` 입니다.

- 아래 데이터를 참조하여 18/19년도에 대해서 Price to Dream Ratio (PDR)을 계산하세요.
  - 변수 이름은 `df2`, feature 이름은 `PDR`로, `df2['PDR']`을 통해 `~` 값이 확인 되어야합니다. 아래의 전제조건을 활용하세요.
  - 점유율 (일반담배 기준) 
    - 18년 : 62%
    - 19년 : 63.5% 
  - 시장의 규모 
     - 18년 : 15조 
     - 19년 : 15조

||구분|종가|발행주식수|시가총액|시장점유율|PDR|
|:-:|:-:|:-:|:-:|:-:|:-:|:-:|
|0|18년|101500|137292497|13935188445500|62.0|~|
|1|19년|93800|137292497|12878036218600|63.5|~|


In [1]:
import pandas as pd
import datetime as dt

url = 'https://ds-lecture-data.s3.ap-northeast-2.amazonaws.com/kt%26g/kt%26g_0.csv'
headers = ['분기', '매출액', '영업이익', '영업이익(발표기준)', '세전계속사업이익',
           '당기순이익', '당기순이익(지배)', '당기순이익(비지배)', '자산총계', '부채총계',
           '자본총계', '자본총계(지배)', '자본총계(비지배)', '자본금', '영업활동현금흐름',
           '투자활동현금흐름', '재무활동현금흐름', '영업이익률', '순이익률', 'ROE(%)',
           'ROA(%)', '부채비율', '자본유보율', 'EPS(원)', 'PER(배)']

df = pd.read_csv(url, header=None)
df.columns = headers

df.dtypes
# 매출액 object, 영업이익 object이므로 int로 형변환
def toInt(string):
  return int(string.replace(',', ''))

# 영업이익률 = (영업이익/매출액) * 100
df['영업이익률2'] = (df['영업이익'].apply(toInt) / df['매출액'].apply(toInt)) * 100
df

분기     매출액   영업이익 영업이익(발표기준)  ...   자본유보율 EPS(원) PER(배)     영업이익률2
0   19/6  12,578  3,991      3,991  ...  803.22  2,351  13.77  31.730005
1   19/9  13,222  3,825      3,825  ...  836.86  2,337  13.87  28.929058
2  19/12  11,982  2,523      2,523  ...  850.82    873  12.42  21.056585
3   20/3  11,784  3,150      3,150  ...  823.16  2,134   9.71  26.731161
4   20/6  13,188  3,947      3,947  ...  837.40  2,135  10.46  29.928723

[5 rows x 26 columns]

In [2]:
headers = ['구분', '종가', '발행주식수', '시가총액', '시장점유율']
data = [['18년', 101500, 137292497, 13935188445500, 62.0],
        ['19년', 93800, 137292497, 12878036218600, 63.5]]

df2 = pd.DataFrame(data)
df2.columns = headers

# PDR = (시가총액 / (시장규모 * 점유율)) * 100
tam = 15000000000000
df2['PDR'] = (df2['시가총액'] / (tam * df2['시장점유율']/100)) * 100
df2

구분      종가      발행주식수            시가총액  시장점유율         PDR
0  18년  101500  137292497  13935188445500   62.0  149.840736
1  19년   93800  137292497  12878036218600   63.5  135.202480

## 2. String Manipulation

- replace가 아닌 다른 방법을 사용하여 19년도 2분기의 매출액 을 integer로 변경하세요. 결과값은 12578이 나와야합니다.
  


In [3]:
import re

def comma_num_to_int(text):
  if str(type(text)) == "<class 'str'>":
    print('in')
    n = re.sub(",", "", text)
    return int(n)
  else:
    print('not in')
    print(type(text))
    return text

target = df.index[(df['분기'] == '19/6')]
df['매출액'][target[0]] = comma_num_to_int(df.loc[target[0]]['매출액'])
df['매출액'][target[0]]

in


/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:14: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


12578

## 3. Apply
- 데이터셋의 분기를 제외한 모든 feature에 대해서 수치형 데이터 (`int`, `float`) 로 수정하고, `describe`를 통해 summary statistics 를 확인하세요.
- 단 결측치는 `0`으로 처리합니다.

In [13]:
import copy

# 각 컬럼별 타입 확인
df.dtypes

# 결측치 처리
df.isna()
df.fillna(0)

for column in df.columns:
  if(df)
  print(column)



분기
매출액
영업이익
영업이익(발표기준)
세전계속사업이익
당기순이익
당기순이익(지배)
당기순이익(비지배)
자산총계
부채총계
자본총계
자본총계(지배)
자본총계(비지배)
자본금
영업활동현금흐름
투자활동현금흐름
재무활동현금흐름
영업이익률
순이익률
ROE(%)
ROA(%)
부채비율
자본유보율
EPS(원)
PER(배)
영업이익률2


## 🔥 도전과제



### NA Value Handling

- 19년도 4분기의 당기순이익(비지배) 부분을 Na로 대체하세요
- 이후 해당 결측치를 mean imputation 방법을 사용하여 처리하세요.



In [ ]:
# 이 곳에 답안을 작성하시길 바랍니다

### Feature Engineering

Relative Perfomance 라는 새로운 feature를 계산하세요.

이는 최근 1년치 매출액의 평균값을 기준으로 

- 10% 이상 -> `S`
- 5% 이상 -> `A`
- -5 ~ 5% -> `B`
- -5%이하 -> `C`
- 10%이하 -> `D`

라는 값을 갖는 feature입니다.

20년도 2분기에 해당하는 결과값은 `A`가 나와야합니다.

각각에 해당하는 등급이 나오기 위해서 필요한 매출액을 추가로 서술하세요.

In [ ]:
# 이 곳에 답안을 작성하시길 바랍니다